src/main/cpp/src/hyper_log_log_plus_plus_host_udf.cu

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "hyper_log_log_plus_plus.hpp"
+#include "hyper_log_log_plus_plus_host_udf.hpp"
+
+#include <cudf/aggregation.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/groupby.hpp>
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+
+namespace spark_rapids_jni {
+
+namespace {
+
+template <typename cudf_aggregation>
+struct hllpp_udf : cudf::host_udf_base {
+  static_assert(std::is_same_v<cudf_aggregation, cudf::reduce_aggregation> ||
+                std::is_same_v<cudf_aggregation, cudf::groupby_aggregation>);
+
+  hllpp_udf(int precision_, bool is_merge_) : precision(precision_), is_merge(is_merge_) {}
+
+  [[nodiscard]] input_data_attributes get_required_data() const override
+  {
+    if constexpr (std::is_same_v<cudf_aggregation, cudf::reduce_aggregation>) {
+      return {reduction_data_attribute::INPUT_VALUES};
+    } else {
+      return {groupby_data_attribute::GROUPED_VALUES,
+              groupby_data_attribute::GROUP_OFFSETS,
+              groupby_data_attribute::GROUP_LABELS};
+    }
+  }
+
+  [[nodiscard]] output_type operator()(host_udf_input const& udf_input,
+                                       rmm::cuda_stream_view stream,
+                                       rmm::device_async_resource_ref mr) const override
+  {
+    if constexpr (std::is_same_v<cudf_aggregation, cudf::reduce_aggregation>) {
```

Comment:
With the new interfaces, this needs to be separated into two separate `operator()` functions.

Reply:
done